nodeJS/express/routing.md

```diff
@@ -0,0 +1,246 @@
+ */
+app.get("/:username/messages/:messageId", (req, res) => {
+  console.log(req.params);
+  res.end();
+});
+```
+
+Now we can easily extract values from the request path for use within our middleware chain!
+
+#### Query parameters
+
+<span id="query-parameters"></span>
+
+Query parameters are a unique and optional part of a URL that appear at the end, starting with a `?`. They are special as they are not actually considered part of the path itself, but are essentially more like arguments we can pass in to a given path. For example, `/odin/messages?sort=date&direction=ascending` will still match the route with the `/:username/messages` path, but we can access the `sort=date` and `direction=ascending` key-value pairs inside the middleware chain.
+
+Express automatically parses any query parameters in a request and will populate the `req.query` object with any key-value pairs it finds. If any keys are repeated, Express will put all values for that key into an array.
+
+```javascript
+/**
+ * GET /odin/messages?sort=date&direction=ascending will log
+ * Params: { username: "odin" }
+ * Query: { sort: "date", direction: "ascending" }
+ *
+ * GET /odin/messages?sort=date&sort=likes&direction=ascending will log
+ * Params: { username: "odin" }
+ * Query: { sort: ["date", "likes"], direction: "ascending" }
+ */
+app.get("/:username/messages", (req, res) => {
+  console.log("Params:", req.params);
+  console.log("Query:", req.query);
+  res.end();
+});
+```
+
+You may have already seen this with websites like YouTube. Every YouTube video is given a code and to watch that video, you navigate to `https://www.youtube.com/watch`, passing that video code as a query parameter with the `v` key. So appending `?v=xm3YgoEiEDc` will [request `/watch` from YouTube using `xm3YgoEiEDc` as the `v` query parameter](https://www.youtube.com/watch?v=xm3YgoEiEDc).
+
+### Routers
+
+So far, we've not been using many routes, and all routes we've shown have been attached to `app`, our server itself. In a real application with lots of routes, we'd probably want to organize our routes into groups and extract each group out to their own file. We could also then more easily write things that affect only the routes in that file, and not any others.
+
+Say we were making a library app and we wanted pages that dealt with books and pages that dealt with authors. That's on top of the homepage and any other miscellaneous pages like "about" or "contact".
+
+We might want our server to handle the following routes:
+
+```text
+GET /
+GET /about
+GET /contact
+POST /contact
+
+GET /books
+GET /books/:bookId
+GET /books/:bookId/reserve
+POST /books/:bookId/reserve
+
+GET /authors
+GET /authors/:authorId
+```
+
+It'd be nice if we could extract the route groups to their own files. It'd also be nice if the server didn't need to check any routes that didn't start with `/authors` if a request was made to a path that started with `/authors`. We can do that using routers!
```

Comment:
> It'd also be nice if the server didn't need to check any routes that didn't start with `/authors` if a request was made to a path that started with `/authors`

^ this sentence is confusing, can you explain what it's trying to say?

Reply:
The original intent is that if you have 30 routes in a `/books` router and a request is for `/authors/*`, it'll straight up skip the routes in the `/books` router.

But thinking about it again, I'm not really sure it's an "issue" worth mentioning. Should be perfectly sufficient to say that a request is passed into the matching router, and that's it.

Reply:
Hmm, not sure if it matters to the learners how many routes the server has to match before it finds the correct one.

Reply:
Makes sense in hindsight - this can be omitted